### PR TITLE
[WIP][distributed] add destructor to process group agent

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28188 [distributed] add destructor to process group agent**

Implements destructor for process group agent, which for now just
detaches the `listenerThread` if it is `joinable` so that we don't get exceptions when the agent
exits without joining. In the usual case where the user calls `rpc.join_rpc()`, this destructor does not do anything (i.e., it behaves as the same default which was implicitly used previously). This is needed to test
https://github.com/pytorch/pytorch/pull/26336

See more context in the issue https://github.com/pytorch/pytorch/issues/28126 and PR https://github.com/pytorch/pytorch/pull/26336.

Differential Revision: [D17969586](https://our.internmc.facebook.com/intern/diff/D17969586/)

### CODE REVIEW

File: torch/csrc/distributed/rpc/process_group_agent.cpp [hunk 1]
```diff
@@ -149,6 +149,15 @@ ProcessGroupAgent::ProcessGroupAgent(
   }
 }
 
+ProcessGroupAgent::~ProcessGroupAgent() {
+  // TODO: when futures handle failures
+  // (https://github.com/pytorch/pytorch/issues/25531), this should mark
+  // existing futures as completed with an exception.
+  if (listenerThread_.joinable()) {
+    listenerThread_.detach();
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: Does this mean either every agent calls join, or none of them do. Otherwise, we could still hit the same error?

Comment (mrshenli) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: Can we add a log/warning here, as we are living one thread alive? Would I be correct if I assume the listener thread will only die when the process dies?

File: torch/csrc/distributed/rpc/process_group_agent.h [hunk 2]
```diff
@@ -40,6 +40,8 @@ class ProcessGroupAgent : public RpcAgent {
       std::shared_ptr<c10d::ProcessGroup> pg,
       int numSendRecvThreads = 4);
 
+  ~ProcessGroupAgent();
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/process_group_agent.h]: override?

File: torch/csrc/distributed/rpc/process_group_agent.cpp (hunk 1 above)
Comment (rohan-varma) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: @mrshenli, which thread are we leaving alive? In this case, we detach the listener thread if it is still joinable. This ensures that the thread object is destructed in a valid manner (i.e. it is joined or detached).

File: torch/csrc/distributed/rpc/process_group_agent.cpp [hunk 3]
```diff
@@ -147,10 +147,14 @@ ProcessGroupAgent::ProcessGroupAgent(
   for (int rank = 0; rank < (int)tmpWorkerIds.size(); ++rank) {
     allWorkerInfo_.emplace_back(std::move(tmpWorkerIds[rank]), rank);
   }
+
+  // construct PythonRpcHandler singleton here
+  PythonRpcHandler::getInstance();
```
Comment (zhaojuanmao) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: @rohan-varma why is this needed? I removed it because I tried to make process group agent independent of python. PythonRpcHandler has dependency on  python interpreter. if process group agent is used in pure C++, this line will not work.

File: torch/csrc/distributed/rpc/process_group_agent.cpp (hunk 1 above)
Comment (mrshenli) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: When does the `listenerThread_` get destructed?

Comment (rohan-varma) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: The data members of process group agent should be destructed after this destructor body runs. So in this case after this code, `~listenerThread` is called and the `listenerThread` is destructed. It is safe to destroy a thread when the thread was joined or detached. Previously, we would enforce that `join` is called (via `ProcessGroupAgent::join`, so there was no need to possibly detach in the destructor. But now, we check if it has been joined, if not, we detach it, so that the `listenerThread` does not throw an exception (https://en.cppreference.com/w/cpp/thread/thread/~thread)

Comment (xush6528) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: THere is a lint message.

```
Avoid detaching threads and letting them run free; there's a good chance they will access destroyed data after main() exits.  Consider not detaching, and later join()'ing the thread.
```

Does this imply that even if the std::thread data structure is destructed, the POSIX thread`pthread_t ` could still be running?

Comment (mrshenli) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: > the POSIX thread pthread_t could still be running?

That's also what I intended to ask. By `listenerThread_`, I actually mean the thread. I am not clear when that thread is destructed, and if we don't feed `recvAnysource` with a message, will that leave it hanging until the process dies? When that happens, will that thread throw any error, and does those errors matter?

File: test/rpc_test.py [hunk 4]
```diff
@@ -858,3 +859,13 @@ def test_func():
...
+
+    def test_process_group_agent_destructor(self):
+        dist.init_process_group(backend="gloo", init_method=self.init_method, timeout=timedelta(seconds=10))
+        rpc.init_model_parallel(
+            self_name="worker%d" % self.rank,
+            backend=TEST_CONFIG.rpc_backend,
+            self_rank=self.rank,
+            init_method=self.init_method,
+        )
+        # exception would be raised here if process group agent exited without detaching/joining threads.
```
Comment (pritamdamania87) [inline test/rpc_test.py]: With your changes, an exception shouldn't be raised right?

File: test/rpc_test.py [hunk 5]
```diff
@@ -858,3 +859,13 @@ def test_func():
 
         if TEST_CONFIG.rpc_backend == RpcBackend.PROCESS_GROUP:
             self.assertEqual(test_func(), "expected result")
+
+    def test_process_group_agent_destructor(self):
+        dist.init_process_group(backend="gloo", init_method=self.init_method, timeout=timedelta(seconds=10))
```
Comment (pritamdamania87) [inline test/rpc_test.py]: Can you enhance with tests by doing some RPCs as well? I think that might trigger some of the issues with condition variables and mutexes.

File: torch/csrc/distributed/rpc/process_group_agent.cpp (hunk 1 above)
Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: We should try to shutdown the listenerThread_ as soon as possible and then call join() on it. We can change `listenLoop` to have `while(!shutdown)` and then in the destructor set shutdown (needs to be atomic bool) to true and then call join() on the thread.

We also need to shutdown `threadPool_` cleanly. The reason is that its not safe for the destructor of futureCV_ to run without calling notify on it (https://en.cppreference.com/w/cpp/thread/condition_variable/~condition_variable). Also, its not safe for the destructor of std::mutex to run if a thread is still holding the lock (https://en.cppreference.com/w/cpp/thread/mutex). Note that there is also a case where a thread calling `send` might be holding `futureMutex_` when the destructor runs (maybe we can move that codeblock inside enqueueSend and run it in the threadpool?). In general, we should try to bring down all the threads in ProcessGroupAgent down as quickly as possible and still in a safe manner.

Also, make sure that your destructor doesn't throw an exception: https://isocpp.org/wiki/faq/exceptions#dtors-shouldnt-throw

File: test/rpc_test.py (hunk 4 above)
Comment (rohan-varma) [reply test/rpc_test.py]: Yep, this comment was meant to indicate what would happen in the test if the process group agent was not destructed correctly. I'll make this comment more clear.

File: torch/csrc/distributed/rpc/process_group_agent.cpp (hunk 1 above)
Comment (xush6528) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: @pritamdamania87 
`listenerThread_` is blocked on calling `pg_->recvAnysource(..)`, there is no good way to make it get back to the loop condition check.

Comment (rohan-varma) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: > @pritamdamania87
> `listenerThread_` is blocked on calling `pg_->recvAnysource(..)`, there is no good way to make it get back to the loop condition check.

As discussed, this is also why we'd want to be able to have a gloo process send a message to itself.

Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: Instead of doing `pg_->recvAnysource(preamble, pg_->getRank())->wait();`, we could do something like this (ideally there should be some way of interrupting pg->recvAnySource):

```
  work = pg_->recvAnysource(preamble, pg_->getRank())
  while (!work.isCompleted() && !shutdown) {
            std::this_thread::sleep_for(
          std::chrono::milliseconds(sleepTime));
  } 
 
 if (shutdown) {
  break;
 }
```

Comment (rohan-varma) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: > Instead of doing `pg_->recvAnysource(preamble, pg_->getRank())->wait();`, we could do something like this (this is ugly and ideally there should be some way of interrupting pg->recvAnySource):
> 
> ```
>   work = pg_->recvAnysource(preamble, pg_->getRank())
>   while (!work.isCompleted() && !shutdown) {
>             std::this_thread::sleep_for(
>           std::chrono::milliseconds(sleepTime));
>   } 
>  
>  if (shutdown) {
>   break;
>  }
> ```

After some testing, I'm not sure if this approach would work. The `wait` that is called in the current code runs `ProcessGroupGloo::RecvWork::wait`, which sets the `completed_` variable that `isCompleted()` needs to be true to return. Without waiting and blocking on `isCompleted`, there is a forever loop.

Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: I was thinking about introducing a ProcessGroupGloo.shutdown method. The shutdown method would do what the destructor does currently. In addition to that it iterates through all contexts in `contexts_` and calls `closeConnections` on each one. Although looking through the gloo code it seems like we hold a lock when we call `read` and `close` needs to hold the same lock, so we might actually end up waiting forever.

**Comment by rohan-varma** [comment]
Closing since this is not the approach we want to take (see https://github.com/facebookincubator/gloo/issues/227 for how we will solve this in gloo itself)
